File: UDPSenderReceiver/Lab2/src/decryption.rs
use crate::encryption::{swap_bits, xor_key, combine_halves};

pub fn decrypt_round(data_block: u16, key: u8) -> u16 {
    let l_data = (data_block >> 8) as u8; //split into left and right halves
    let r_data = (data_block & 0xFF) as u8;
    
    let r_data_decrypted = xor_key(r_data, key); //xor (to undo encryption xor)
    let combined = combine_halves(l_data, r_data_decrypted); //recombine 
    
    swap_bits(combined, 0, 8, 8) //swap halves back to get original 16 bit data block
}

File: SymmetricEncryptionSimpleDES/PythonImplementation/decryption.py
from encryption import swapBits, xor_key, combine_halves

def decrypt_round(data_block, key):
    # 1. Split into halves first
    l_data = (data_block >> 8) & 0xFF
    r_data = data_block & 0xFF
    # 2. Reverse the XOR (XORing with the same key undoes it)
    r_data = xor_key(r_data, key)
    # 3. Recombine to prepare for the bit swap
    data_block = combine_halves(l_data, r_data)
    # 4. Swap the 8-bit halves back (0 and 8 are the same positions used in encrypt)
    # This puts the original left side back on the left and right on the right
    original_block = swapBits(data_block, 0, 8, 8)
    return original_block

if __name__ == "__main__":
    # Test case using the result from your encryption
    encrypted_val = 0b1100110101011011 
    key = 0b11110000
    decrypted = decrypt_round(encrypted_val, key)
    print(f"Encrypted: {encrypted_val:016b}")
    print(f"Decrypted: {decrypted:016b}")